hydro_deploy/core/src/localhost/launched_binary.rs

```diff
@@ -186,7 +250,107 @@ impl LaunchedBinary for LaunchedLocalhostBinary {
+                                                .into_iter()
+                                                .rev()
+                                                .map(|inline| {
+                                                    inline
+                                                        .function
+                                                        .unwrap_or_else(|| "unknown".to_string())
+                                                })
+                                                .join(";"),
+                                        );
+                                    } else {
+                                        frame_lookuped.push(lookuped.symbol.name);
+                                    }
+                                } else {
+                                    frame_lookuped.push("unknown".to_string());
+                                }
+                            } else {
+                                frame_lookuped.push("unknown".to_string());
+                            }
+                        }
+
+                        let all_leaves_grouped = thread
+                            .samples
+                            .stack
+                            .iter()
+                            .enumerate()
+                            .filter_map(|(idx, s)| s.map(|s| (idx, s)))
+                            .map(|(idx, leaf)| (leaf, thread.samples.weight[idx]))
+                            .chunk_by(|v| v.0)
+                            .into_iter()
+                            .map(|(leaf, group)| {
+                                let weight = group.map(|t| t.1).sum();
+                                (leaf, weight)
+                            })
+                            .collect::<Vec<(usize, u64)>>();
+
+                        for (leaf, weight) in all_leaves_grouped {
+                            let mut cur_stack = Some(leaf);
+                            let mut stack = vec![];
+                            while let Some(sample) = cur_stack {
+                                let frame_id = thread.stack_table.frame[sample];
+                                stack.push(frame_lookuped[frame_id].clone());
+                                cur_stack = thread.stack_table.prefix[sample];
+                            }
+
+                            *folded_frames.entry(stack).or_default() += weight;
+                        }
+                    }
+
+                    let mut output = String::new();
+                    for (stack, weight) in folded_frames {
+                        for (i, s) in stack.iter().rev().enumerate() {
+                            if i != 0 {
+                                output.push(';');
+                            }
+                            output.push_str(s);
+                        }
+
+                        output.push_str(&format!(" {}\n", weight));
+                    }
+
+                    output.into()
```

Comment:
Could we wrap this up in a helper function, put it and the helper structs into a separate module for samply stuff?